File: cs/Problems/Day4.cs
namespace AOC2023.Problems;

public class Day4 : IProblem<int>
{
    public int Solve(ReadOnlySpan<char> input) => CalculateScratchCardPoints(input);

    private static int CalculateScratchCardPoints(ReadOnlySpan<char> input)
    {
        int cardPoints = 0;
        Span<Range> gameLines = stackalloc Range[200];

        int count = input.Split(gameLines, InputReader.NewLine);
        for (int i = 0; i < count; i++)
        {
            var game = input[gameLines[i]];
            var numbers = game[(game.IndexOf(":") + 1)..];

            var separator = numbers.IndexOf('|');
            var foundSlice = numbers[(separator + 1)..];
            var winnersSlice = numbers[..(separator - 1)];

            int gamePoints = 0;

            for (int j = 0; j < foundSlice.Length; j += 3)
            {
                bool exists = winnersSlice.Contains(foundSlice[j..(j + 3)], StringComparison.Ordinal);
                gamePoints = exists ? Math.Max(1, gamePoints * 2) : gamePoints;
            }

            cardPoints += gamePoints;
        }

        return cardPoints;
    }
}

File: go/problems/day4.go
package problems

import (
	"strings"
)

type Day4 struct{}

func (d Day4) Solve(input string) int {
	return calculateScratchCardPoints(input)
}

func calculateScratchCardPoints(input string) int {
	cardPoints := 0

	for _, game := range strings.Split(input, NewLine) {
		gameSep := strings.Index(game, ":")
		numbSep := strings.Index(game, "|")

		found := game[gameSep+1 : numbSep-1]
		winners := game[numbSep+1:]

		gamePoints := 0

		for j := 0; j < len(found); j += 3 {
			if strings.Contains(winners, found[j:j+3]) {
				gamePoints = max(1, gamePoints*2)

			}
		}

		cardPoints += gamePoints
	}

	return cardPoints
}
